### PR TITLE
✨ Adds consistent redirects handling for cf workers.

### PR DESCRIPTION
**Pull request by danyx23**
Cloudflare has support for redirects but only up to a limited number (I think 2000?). We have a lot of chart redirects in the chart_redirects table (about 1600 or so). Because everything under the /grapher/ route on our website has to run through a CF function anyways (to update the thumbnail image to include the correct query string), we decided to bake these redirects into a json file and do custom redirects handling in the CF function. 

We did this for html pages already but not for thumbnails, but now that we also have `.config.json` requests to worry about and would like to have consistent redirects handling that doesn't overly complicate the logic of our handlers.

This PR makes it so that both requests for the html pages at `/grapher/[slug]` and requests for `/grapher/[slug].config.json` will be redirected according to the _grapherRedirects.json file using a catch error handler. This is done as a catch handler that checks for 404 pages so that the common, happy path does not have to fetch the redirects file.

### CODE REVIEW

**Comment by danyx23** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3895?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#3884** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3884?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#3880** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3880?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#3879** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3879?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#3895** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3895?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#3870** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3870?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#3871](https://github.com/owid/owid-grapher/pull/3871) <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3871?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#3867** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3867?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#3865](https://github.com/owid/owid-grapher/pull/3865) <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3865?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#3827** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3827?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#3847](https://github.com/owid/owid-grapher/pull/3847) <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3847?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#3782** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3782?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2 other dependent PRs ([#3793](https://github.com/owid/owid-grapher/pull/3793) <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3793?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#3890](https://github.com/owid/owid-grapher/pull/3890) <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3890?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#3767** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3767?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#3743](https://github.com/owid/owid-grapher/pull/3743) <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3743?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#3846** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3846?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @danyx23 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: functions/grapher/[slug].ts [hunk 1]
```diff
@@ -27,10 +24,16 @@ router
...
+        headers: {
+            "Access-Control-Allow-Origin": "*",
+        },
+    })
+}
+
 export const onRequestGet: PagesFunction = async (context) => {
-    // Makes it so that if there's an error, we will just deliver the original page before the HTML rewrite.
-    // Only caveat is that redirects will not be taken into account for some reason; but on the other hand the worker is so simple that it's unlikely to fail.
-    context.passThroughOnException()
```
Comment (marcelgerber) [inline functions/grapher/[slug].ts:33]: We should really try to keep this in as the first line of business.

It allows us that if any unhandled exception happens in this handler, we're still _failing open_ and are sending over the unmodified `/grapher/...` response as-is from the origin server. Which is a good fallback (and will work for graphers unless there's a redirect - but never for `.json`, `.png` and `.svg`, obviously.)

See https://developers.cloudflare.com/workers/runtime-apis/context/#passthroughonexception.

File: functions/grapher/[slug].ts [hunk 2]
```diff
@@ -41,7 +44,49 @@ export const onRequestGet: PagesFunction = async (context) => {
...
+            if (e instanceof StatusError && e.status === 404) {
+                console.log("Handling 404 for ", url.pathname)
+                const fullslug = url.pathname.split("/").pop() as string
+
+                const allExtensions = Object.values(extensions)
+                    .map((ext) => ext.replace(".", "\\.")) // for the regex make sure we match only a single dot, not any character
+                    .join("|")
+                const regexForKnownExtensions = new RegExp(
+                    `^(?<slug>.*?)(?<extension>${allExtensions})?$`
+                )
```
Comment (marcelgerber) [inline functions/grapher/[slug].ts:61]: I think this code will be simpler by assuming that a slug can't contain a dot, i.e. something like `/^(?<slug>[^\s\.]+)(?<extension>\.\S+)?$/`.

In theory we could then check after the fact that the extension is valid (and throw a fitting error), but it's also fine if we're not doing that.

File: functions/grapher/[slug].ts [hunk 3]
```diff
@@ -41,7 +44,49 @@ export const onRequestGet: PagesFunction = async (context) => {
...
+
+                const matchResult = fullslug.match(regexForKnownExtensions)
+                const slug = matchResult?.groups?.slug ?? fullslug
+                const extension = matchResult?.groups?.extension ?? ""
```
Comment (marcelgerber) [inline functions/grapher/[slug].ts:65]: So, what this code doesn't currently seem to handle is non-lowercase slugs.
Our rule is that a slug needs to be all-lowercase, and if it isn't we want to redirect it.

In the old code, I tried to be clever about it so we never need to run through two redirects, avoiding the scenario `REDIR-SOURCE -> redir-source -> redir-target`. I don't think we need to take care of that again, which means we would just do:

```ts
if (slug.toLowerCase() !== slug)
  return redir(302, slug.toLowerCase())
```

File: functions/_common/grapherRenderer.ts [hunk 4]
```diff
@@ -215,6 +217,13 @@ export async function fetchGrapherConfig(
 ): Promise<FetchGrapherConfigResult> {
     const fetchResponse = await fetchUnparsedGrapherConfig(slug, env, etag)
 
+    if (fetchResponse.status === 404) {
+        // we throw 404 errors instad of returning a 404 response so that the router
```
Comment (marcelgerber) [inline functions/_common/grapherRenderer.ts:221]:
```suggestion
        // we throw 404 errors instead of returning a 404 response so that the router
```

**Review by marcelgerber — COMMENTED** [review]
Nicely done!

I did find a bunch more cases that we need to cover, or may want to cover in a nicer way.

File: functions/grapher/[slug].ts (hunk 2 above)
Comment (danyx23) [reply functions/grapher/[slug].ts:61]: Yeah true but I think I like the idea of using a list of known extensions as a communication device that is added to further up in the stack. I'm inclined to leave it as is for now.

**Comment by danyx23** [comment]
### Merge activity

* **Sep 10, 7:13 AM EDT**: @danyx23 started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/owid/owid-grapher/3895).

**Pull request merged**